### ISSUE TITLE
The IETF draft has expired last month

### ISSUE THREAD

**Issue by Jopie64** [wamp-proto/wamp-proto#249]
I'm not sure whether this is a problem, but while I was consuling the specs, I noticed that it has expired last month. See [here](https://github.com/wamp-proto/wamp-proto/blob/master/rfc/draft-oberstet-hybi-tavendo-wamp.html#L545)

https://github.com/wamp-proto/wamp-proto/blob/master/rfc/draft-oberstet-hybi-tavendo-wamp.html#L545

**Comment by oberstet** [comment]
@Jopie64 yeah, I am aware. The reason is that the Hybi IETF working group has closed, and hence we need to resubmit the RFC draft as a "personal submission" until we have found a new "host".

**Comment by nigelmegitt** [comment]
+1 to resubmitting this - ratifying through IETF would allow external open standards to reference it.

**Comment by oberstet** [comment]
@nigelmegitt Absolutely. Also: I am still convinced that the IETF would be the best fit for formal standardization - I totally grok the mantra "rough consensus, working code". It'll be a lot of work still. On the good side: I think the "consensus" part should work out well, as we (WAMP implementors) have had a lot of discussions already mastered;)

**Comment by waterbug** [comment]
I'm late to the party, having just discovered that the RFC had expired, but I am also +1 on resubmitting.  I am into development of an application that uses WAMP (crossbar and autobahn-python), and I'd like to be able to cite WAMP as an RFC ... partly because I am in government (NASA) so we like standards ;) and I may be collaborating with NIST (U.S. govt. standards agency).  I also strongly agree with you about "rough consensus, working code" -- the standards world could definitely benefit by more of that!

**Comment by oberstet** [comment]
@waterbug It is fascinating to see even very large organization nowadays being more and more open to community driven, grass-roots technologies. NASA/NIST .. awesome;)

> We really need an updated users list - I _think_ I remember DLR is also using WAMP/Crossbar.io .. for some ground station related stuff .. not exactly built into rockets or safety critical control systems or such.

Anyway, I guess "it just works" might be a factor - we, as WAMP community, have worked hard to make it happen - and there is still some polish here and there to do, though the fundament is stable and proven (after the WAMPv1 => v2 step). 

---

Alright. Please let me give you a little context rgd IETF and the RFC.

The WebSocket protocol was developed in the IETF Hybi WG (where we participated). But the WG has reached it's original charter goals (https://tools.ietf.org/wg/hybi/charters), and hence has been put into "maintenance mode".

It is possible (any time) to place _personal submissions_ to the IETF - however, this cannot get onto a standards track.

To get on the standards track, we need a hosting WG (Hybi would be possible) _and_ an IETF chair overseeing. Rgd the latter I have an informal promise by Salvatore Loreto (https://www.linkedin.com/in/salvatoreloreto/) he'd be willing to fill this role for WAMP, **given that we can organize and demonstrate enough interest**. How exactly this is to be demonstrated is fuzzy/open ..

Now, I figure that interest by NASA/NIST should qualify;)

So if you want to help make it happen: what you could do is to mail the Hybi list (and/or Salvatore), and disclose your NASA/NIST affiliation and interest there.

> I might be able to convince other larger org users to do the same when I can point them to this ..

---

The other thing is: it will be lots of work still. The IETF has high standards, and a lot of very smart people. WebSocket was quite exhausting .. hundreds and hundreds of mails and discussions. Nearly 2 years work. For WAMP, I guess it should be possible to do it in under one year. (WebSocket was obviously a high-profile thing with lots of big guys with stakes).

And my (personal) problem: we are a small startup, short in time, limited resources, and we have bills to pay. So any possible financial support would of course also speed things up;)

For us as a startup, we have projects and PoCs going with companies that already are convinced of WAMP (and Crossbar.io in our case) - they don't care that much about an RFC - "it just works". They won't pay for paperwork;)

If we had the money, would be a no-brainer for me: do an RFC full time.

AND: develop an automated fuzzing testsuite for WAMP - we did that for WebSocket (http://autobahn.ws/testsuite/), virtually used all over the industry now, and technically that paid off big I'd say - the conformance and interoperability of WebSocket implementations is very very good, and I shamelessly take some credit for that;)

Cheers,
/Tobias

**Comment by waterbug** [comment]
@oberstet Thanks for your very thorough and in-depth response!  First, I should correct an impression that I didn't mean to give:  I do work for NASA, but the application I am developing is currently internal R&D, and I strongly suspect I am the only person in NASA who is attempting to apply WAMP ... I am known for taking risks and pushing the envelope, so the fact that I am using crossbar/autobahn should not be taken as "NASA/NIST are interested in WAMP", but rather that WAMP has a single advocate in those organizations -- me (I was at a standards meeting last week and described WAMP to a NIST colleague, but he had never heard of it -- I will try to remedy that).  OTOH, if my little project succeeds, it *could* have big implications -- at least I like to think so. ;)

I should also say that the lack of an RFC is not a show-stopper for me, by any means.  What you say about the cost of pursuing the RFC rings very true -- as I mentioned, I am and have been involved in standardization activities (mainly ISO), and they are extremely cumbersome and expensive in terms of both time and resources -- there has to be a clear and compelling business case to invest that much. IETF is probably somewhat less cumbersome than ISO, but that just means slightly less than ridiculous. ;)

I am working in the area of "collaborative engineering", and at the meeting last week I discovered a new standards initiative called MoSSEC (http://www.mossec.org/), which is designed to apply web services in the engineering tools domain.  It is currently an ISO "New Work Item", and some its initiators were at the meeting.  Currently their prototyping has been with SOAP (ugh, no!) and REST services (ok), and I suggested WAMP as a more powerful alternative -- this was the first time they had heard of it and they kind of groaned at the thought of another implementation method, but I hope to create a test bed using my own application as a starting point.  If that succeeds, it would at least give WAMP an interesting success story from a domain outside of IoT (but there could be IoT connections there as well, since this is the manufacturing industry).  Some more food for thought.

**Comment by oberstet** [comment]
Ahh, I see. NP, I was probably too intrigued by the idea of NASA actually using WAMP and got carried away;) But as you say, who knows ..

> I am known for taking risks and pushing the envelope

+1 - I love that attitude. I'm deeply convinced, it's required to move stuff really forward and make it happen. Of course there are limits, needs a balance.

---

I had a quick glance at the MoSSEC site .. model driven engineering in globally distributed teams I figure. Interestingly, not with a central model/artifact database, but distributed it seems (the DB).

Anyway, I have a rough idea what this is probably about (in the automotive sector, there is AUTOSAR and all kinds of model driven XYZ - or IBM Rational and those heavy tools, which I had exposure to).

I also now understand "love standards" .. hehe .. this is ISO driven initiative? Which also explains the "love" for SOAP in particular (an ISO thing https://www.iso.org/standard/58345.html) and XML. Shudder;)

Even REST will look lame, because it is not an ISO standard. Who cares that it works, whereas SOAP is broken  / BS in practice. Same for XML. Simply sucks. Well. I have strong opinions there.

FWIW, Crossbar.io supports also HTTP/REST. And WAMP of course does not care if some XML is transferred as payload. And of course, I'd say "distributed teams working together on a shared model" could profit from modern, real-time enabled user interfaces etc (-> WAMP).

However, what definitely won't happen: SOAP in Crossbar.io, or XML as a serialization format in WAMP. Both are just too broken / have zero gain from my PoV. Enough rambling .. blood pressure rising when I only think about it;) Ping me if I can be of any further help or with news!

Cheers,
/Tobias

**Comment by oberstet** [comment]
Ah, not to leave a false impression: I do think there are really great ISO standards: C++ comes to mind. An incredible effort. And compilers nowerdays are actually very close. And, FWIW, C++ 20 will have co-routines - jay!

**Comment by oberstet** [comment]
Alright, I have done cleanups, generated a new RFC version.

The [IETF submission tool](https://datatracker.ietf.org/submit/) seems to be in maintenance currently .. I cannot submit it:

> The I-D submission tool will be reopened after 2017-03-26 23:59 CDT (IETF-meeting local time). 

Anyway, I've deployed to our Web site:

* http://wamp-proto.org/static/rfc/draft-oberstet-hybi-crossbar-wamp.html
* http://wamp-proto.org/static/rfc/draft-oberstet-hybi-crossbar-wamp.txt

---

Surprisingly the spec is at nearly 140 pages! Expect 250 or more when we fill in the gaps. So this is getting quite largish ..

**Comment by waterbug** [comment]
That's great!  I will have a look.  BTW, I share your opinion about SOAP (ugh, broken), and I avoid XML as much as possible -- it just gets in the way.

**Comment by cnlpete** [comment]
It expired again. What is the current status? Do you plan on resubmitting an updated version at some point?

BTW. the IETF-link on http://wamp-proto.org/spec/ semms broken ;)

**Comment by oberstet** [comment]
so I've been going through all open issues, assigning the ones that could go into the next revision here: https://github.com/wamp-proto/wamp-proto/milestone/2

**Comment by ecorm** [comment]
Could we get WAMP-SCRAM in that list? How soon did you want to resubmit?

**Comment by oberstet** [comment]
@ecorm so I collected the list with bugs, and gaps for stuff already on the street in mind .. WAMP-SCRAM is yet to be defined feature. but if you contribute text, why not include it in this milestone? that being said, I really think it would be good to consolidate and polish what we already have - and then come back to defining more;)

**Comment by ecorm** [comment]
@oberstet Now that I think about it, perhaps someone should implement a proof-of-concept implementation to work out the kinks in WAMP-SCRAM before it is submitted as part of the RFC.

**Comment by oberstet** [comment]
yeah, probably best to co-write the text and a prototype implementation - at least that seems to fit the IETF mantra: "rough consensus, working code" =)

**Comment by ecorm** [comment]
Since CppWAMP is client-only, I leave to others to write a prototype implementation.

**Comment by oberstet** [comment]
FWIW, leaving a note here that we use the `RFC` label on issues to track those that have relevance towards getting a RFC shipped: https://github.com/wamp-proto/wamp-proto/labels/RFC

We have nailed 18 .. and 30 are open .. well. It's a start I'd say to finally bring this spec on a proper RFC track ..
